oeps/oep-0003.rst

```diff
@@ -0,0 +1,144 @@
+===================================
+OEP-3: Asynchronous Task Management
+===================================
+
++---------------+-------------------------------------------+
+| OEP           | :doc:`oep-0003`                              |
++---------------+-------------------------------------------+
+| Title         | Asynchronous Task Management              |
++---------------+-------------------------------------------+
+| Last-Modified | 2016-06-13                                |
++---------------+-------------------------------------------+
+| Author        | Jeremy Bowman <jbowman@edx.org>           |
++---------------+-------------------------------------------+
+| Arbiter       |                                           |
++---------------+-------------------------------------------+
+| Status        | Draft                                     |
++---------------+-------------------------------------------+
+| Type          | Architecture                              |
++---------------+-------------------------------------------+
+| Created       | 2016-06-09                                |
++---------------+-------------------------------------------+
+
+Abstract
+========
+
+Proposes a system which allows users to initiate, monitor, and retrieve the
+results of various long-running tasks in a consistent manner.  This system
+leverages existing libraries where practical and makes the development of new
+asynchronous tasks straightforward.
+
+
+Specification
+=============
+
+A new open source Django application will be written to support a user
+interface for managing celery tasks in a consistent manner across services.
+The main goals of the application are, roughly in order:
+
+1. Allow a long-running task triggered by a user to be run asynchronously,
+   while it is clear to the user that they are free to do other work on the
+   site in the meantime
+2. Allow the user to fetch any output of the task after it finishes
+3. Notify the user when the task is complete
+4. Give the user some indication of how far along an in-progress task is
+5. Support prevention of multiple identical tasks from executing concurrently
+
+Celery will be used to run the asynchronous tasks, but the design should be
+capable of adapting to a different task runner later if necessary.  Database
+models will be used to represent tasks as well as their input, output, and
+current status.  It will be possible to present a fan-out of Celery tasks
+instantiated to perform work in parallel as a single task from the user's
+perspective with a meaningful progress indicator.
+
+For notifications, email is sufficient for the initial implementation (but
+other push notifications may be added later).  A task status page for each
```

Comment:
@macdiesel I think @jmbowman's suggestion is reasonable as stated: start with e-mail, combined with a single-task status page. The e-mail only acts as an optional heads up that the single-task status page has new information, which the user could also find by just leaving the status page open and periodically refreshing. That seems like a very useful first cut, with future iterations potentially improving the user experience with a "show all my jobs" dashboard and real-time updating/progress bars and such.

Because e-mail is not technically required for a user to do their job here, I think the "out of date email" and "spam filter" use cases are probably just edge cases we leave out for now and possibly revisit them later. There are far more places in our system where we expect that a user's email is up to date and not sending system mail to spam. This could just be my cynicism but fixing the email issues feels like a hard problem to solve with a low return on that investment.

All that said, I don't think e-mail _by itself_ is sufficient for the first iteration, we really ought to have that single-task status page too.